packages/keyring-eth-qr/src/qr-keyring-v2.ts

```diff
@@ -0,0 +1,228 @@
+  EthMethod,
+  EthScope,
+  type KeyringAccount,
+  KeyringAccountEntropyTypeOption,
+  type KeyringCapabilities,
+  type KeyringV2,
+  KeyringType,
+  type EntropySourceId,
+} from '@metamask/keyring-api';
+import type { AccountId } from '@metamask/keyring-utils';
+import type { Hex } from '@metamask/utils';
+
+import { DeviceMode } from './device';
+import type { QrKeyring } from './qr-keyring';
+
+/**
+ * Methods supported by QR keyring EOA accounts.
+ * QR keyrings support a subset of signing methods (no encryption, app keys, or EIP-7702).
+ */
+const QR_KEYRING_METHODS = [
+  EthMethod.SignTransaction,
+  EthMethod.PersonalSign,
+  EthMethod.SignTypedDataV4,
+];
+
+const qrKeyringV2Capabilities: KeyringCapabilities = {
+  scopes: [EthScope.Eoa],
+  bip44: {
+    deriveIndex: true,
+    derivePath: false,
+    discover: false,
+  },
+};
+
+/**
+ * Concrete {@link KeyringV2} adapter for {@link QrKeyring}.
+ *
+ * This wrapper exposes the accounts and signing capabilities of the legacy
+ * QR keyring via the unified V2 interface.
+ *
+ * All QR keyring accounts are BIP-44 derived (both HD and Account modes use
+ * derivation paths from the hardware device).
+ */
+export type QrKeyringV2Options = {
+  legacyKeyring: QrKeyring;
+  entropySource: EntropySourceId;
+};
+
+export class QrKeyringV2
+  extends EthKeyringWrapper<QrKeyring, Bip44Account<KeyringAccount>>
+  implements KeyringV2
+{
+  readonly entropySource: EntropySourceId;
+
+  constructor(options: QrKeyringV2Options) {
+    super({
+      type: KeyringType.Qr,
+      inner: options.legacyKeyring,
+      capabilities: qrKeyringV2Capabilities,
+    });
+    this.entropySource = options.entropySource;
+  }
+
+  /**
+   * Get the device state from the inner keyring.
+   *
+   * @returns The device state, or undefined if no device is paired.
+   */
+  async #getDeviceState(): Promise<
+    { mode: DeviceMode; indexes: Record<Hex, number> } | undefined
+  > {
+    const state = await this.inner.serialize();
+
+    if (!state?.initialized) {
+      return undefined;
+    }
+
+    return {
+      mode: state.keyringMode,
+      indexes: state.indexes,
+    };
+  }
+
+  /**
+   * Creates a Bip44Account object for the given address.
+   *
+   * @param address - The account address.
+   * @param addressIndex - The account index in the derivation path.
+   * @returns The created Bip44Account.
+   */
+  #createAccount(
+    address: Hex,
+    addressIndex: number,
+  ): Bip44Account<KeyringAccount> {
+    const id = this.registry.register(address);
+
+    const account: Bip44Account<KeyringAccount> = {
+      id,
+      type: EthAccountType.Eoa,
+      address,
+      scopes: [...this.capabilities.scopes],
+      methods: [...QR_KEYRING_METHODS],
+      options: {
+        entropy: {
+          type: KeyringAccountEntropyTypeOption.Mnemonic,
+          id: this.entropySource,
+          groupIndex: addressIndex,
+          derivationPath: this.inner.getPathFromAddress(address) ?? '',
```

Comment:
We might want to treat this as an error actually. If we cannot get this info from the device, then that probably means we cannot use the device also to derive the proper address? 🤔

IMO, the derivation path should not be empty